src/Discord.Net.WebSocket/DiscordSocketClient.cs

```diff
@@ -1355,24 +1356,35 @@ public Task DownloadUsersAsync(params IGuild[] guilds)
                                             return;
                                         }
 
-                                        SocketPresence before;
+                                        SocketPresence beforePresence;
+                                        SocketGuildUser beforeGuild;
+                                        SocketGlobalUser beforeGlobal;
                                         var user = guild.GetUser(data.User.Id);
                                         if (user != null)
                                         {
-                                            before = user.Presence.Clone();
+
+                                            beforePresence = user.Presence.Clone();
+                                            beforeGlobal = user.GlobalUser.Clone();
+                                            beforeGuild = user.Clone();
                                             user.Update(State, data);
                                         }
                                         else
                                         {
-                                            before = new SocketPresence(UserStatus.Offline, null);
+                                            beforePresence = new SocketPresence(UserStatus.Offline, null);
                                             user = guild.AddOrUpdateUser(data);
+                                            beforeGlobal = user.GlobalUser.Clone();
+                                            beforeGuild = user.Clone();
+                                        }
+
+                                        await _userPresenceUpdatedEvent.InvokeAsync(guild, user, beforePresence, user.Presence).ConfigureAwait(false);
```

Comment:
Are there cases where presence wouldn't have changed in this event? Username change? We should avoid raising the event if there are no changes.